### PR TITLE
Suggest types - fixes #1446

### LINKED ISSUE #1446: Improve Error Reporting: Type suggestions not shown consistently
Related to #1102

The following code snippet shows that different error messages are shown depending on how to express the value. The first one shows a simple error message; the second one shows suggestions on valid types.

```fsharp
let data = System.Collections.Generic.Lst()
let dataTwo : System.Collections.Generic.Lst = ResizeArray()
```

CC: @forki

### PR DESCRIPTION
**Pull request by forki**
image

### CODE REVIEW

**Comment by KevinRansom** [comment]
@forki 
Test errors:
```
+++ Warnings (SuggestTypesInModule.fs) +++
Microsoft (R) F# Compiler version 4.1
Copyright (c) Microsoft Corporation. All Rights Reserved.
(6,36): error FS0039: The type 'Lst' is not defined in 'System.Collections.Generic'.

Maybe you want one of the following:

   List

   IList

   Comparer

   Dictionary

   IComparer

*** The following necessary lines were never matched:
***	error FS0039:.+The value, constructor, namespace or type 'Lst' is not defined


*** The following necessary lines were incorrectly matched:

Unexpected Compiler Output 
FAIL
``

**Comment by forki** [comment]
Yep I'm on it

**Comment by forki** [comment]
@enricosada @KevinRansom this is a bit tricky. the build doesn't come to neg2 test on my machine because other stuff is failing.

So I can't really check what's wrong.

File: src/fsharp/NameResolution.fs [hunk 1]
```diff
@@ -2014,8 +2014,15 @@ let rec ResolveExprLongIdentInModuleOrNamespace (ncenv:NameResolver) nenv (typeN
                 else 
                     NoResultsOrUsefulErrors
 
-            AtMostOneResult id.idRange ( tyconSearch +++   moduleSearch +++ raze (UndefinedName(depth,FSComp.SR.undefinedNameValueConstructorNamespaceOrType,id,NoPredictions)))
-
+            match tyconSearch +++ moduleSearch with
```
Comment (KevinRansom) [inline src/fsharp/NameResolution.fs]: @forki 
The AtMostResult indicates that this error can produce a message more than once:

Indeed the error messages produced in the failing neg2 testcase looks like this:

````

neg2.fsx(3,59,3,66): typecheck error FS3033: The type provider 'Provider.GoodProviderForNegativeTypeTests1' reported an error: Kaboom

neg2.fsx(3,59,3,66): typecheck error FS3033: The type provider 'Provider.GoodProviderForNegativeTypeTests1' reported an error: Kaboom

neg2.fsx(3,59,3,66): typecheck error FS0039: The value, constructor, namespace or type 'TheHype' is not defined

Maybe you want one of the following:

   TheType

````

Comment (forki) [inline src/fsharp/NameResolution.fs]: AtMostOneResult was moved to line 2025. but thanks for the test report. I will try to fix that.

**Comment by forki** [comment]
ok this is ready for review

**Comment by KevinRansom** [comment]
@forki 
Okay looks good to me.  Pulling ... 

Thanks for taking care of this :

The text "Maybe you want pne of the following" seems hokey to me.  Could we consider changing it to something more like "Consider one of the following:"

Kevin

**Comment by forki** [comment]
Yes we should change that. I'm not native speaker so I hope others find a
good solution

Am 23.08.2016 10:12 nachm. schrieb "Kevin Ransom (msft)" <
notifications@github.com>:

**Pull request merged**